pkg/validators/middleware.go

```diff
@@ -0,0 +1,33 @@
+package validators
+
+import (
+	"github.com/mt-sre/addon-metadata-operator/pkg/types"
+)
+
+// Middleware abstracts behavior which extends types.ValidatorRunner
+// functionality by executing additional steps before/after validations
+// are run such as adding retries or logging.
+type Middleware interface {
+	// Wrap applies the implemented middleware to an instance of types.ValidatorRunner
+	Wrap(types.ValidatorRunner) types.ValidatorRunner
```

Comment:
An issue with this design is if we need to access `Validator` fields, for example Code or Name.

Say we want to write a `PrometheusMiddleware`, to register results, execution time and other metrics, per validator. How do we access `v.Code` or `v.Name` to add specific labels?

We would have to cast the `types.ValidatorRunner` to `types.Validator` everytime.

Reply:
Yes definitely, there are two options:

1. We could follow the pattern of injecting dependencies when initializing the MW like [so](https://github.com/mt-sre/addon-metadata-operator/pull/63/files#diff-5ca607962a1e1b55993f9c37ade457ffa9aa0a8f7b55dacf5a186f597680dd98R18-R20)
2. Or we could do what you're suggesting which is a little more complex where we pass around ValidatorParams with every `ValidateFunc` like `type ValidateFunc func(params ValidatorParams, mb MetaBundle) ValidatorResult`

I think realistically we need to do the second because then we can actually pass a `context.Context`, logger, clients, etc... to the `ValidateFunc`'s themselves and then middleware will just access those params in it's closure.

Reply:
Implemented the second option with latest commit

Reply:
I like the second options better. We have tons of flexibility to define what this `cfg` should contain too.